openapi2kong should allow multiple "foreign key" plugins

In Konnect some of our "gateway services" use many instances of the rate-limiting-advanced plugin and have multiple "foreign keyed" resources. This is a helpful architecture for doing things like rate limiting tiers for a single service. 

Example would be an RLA that is "scoped" to: consumerA + serviceA. Another instance of an RLA is "scoped" to: consumerB + serviceA. There does not seem to be a way to do this using openapi2kong. It seems to try to "promote" both and the second one fails. This issue seems to happen even with 2 different plugins that have consumers.

To replicate add two RLA plugins that are configured with "consumers" to two paths in an openApi file.

Simply uncommenting the consumer on 1 of the 2 allows the command to succeed. I'm using:

deck 1.38 on a Mac. The command issued is pretty standard:  
  deck file openapi2kong -s <openApi-file.yaml> -o <output-file.yaml> --inso-compatible

FWIW - it would be great if "root" level plugins could be put into the openApi file using some "x-kong-plugins:" or similar. The final structure in deck for these types of configurations look like:

 ...
plugins: <-- at the root of the deck file
- config:
     limit:
     - 5
     ...
   consumer: consumerA
   service: serviceA
   name: rate-limiting-advanced
- config:
     limit:
     - 10
     ...
   consumer: consumerB
   service: serviceA
   name: rate-limiting-advanced


Error produced:
panic: interface conversion: interface {} is nil, not string

goroutine 1 [running]:
github.com/kong/go-apiops/openapi2kong.Convert.func1(0x1000100?, 0x14bf62588?)
        github.com/kong/go-apiops@v0.1.33/openapi2kong/openapi2kong.go:1126 +0x1d0
sort.insertionSort_func({0x14000123a90?, 0x140003ae080?}, 0x0, 0x3)
        sort/zsortfunc.go:12 +0xc0
sort.pdqsort_func({0x14000123a90?, 0x140003ae080?}, 0x140001975a8?, 0x0?, 0x0?)
        sort/zsortfunc.go:73 +0x260
sort.Slice({0x103d9db20?, 0x140001975a8?}, 0x14000123a90)
        sort/slice.go:29 +0xbc
github.com/kong/go-apiops/openapi2kong.Convert({0x14000174000, 0x769e, 0x769f}, {{0x0, 0x0, 0x0}, {0x0, 0x0}, {0x6b, 0xa7, ...}, ...})
        github.com/kong/go-apiops@v0.1.33/openapi2kong/openapi2kong.go:1122 +0x1a18
github.com/kong/deck/cmd.executeOpenapi2Kong(0x140001bcb00?, {0x1400010bef0?, 0x4?, 0x103a49a2a?})
        github.com/kong/deck/cmd/file_openapi2kong.go:61 +0x3f4
github.com/spf13/cobra.(*Command).execute(0x14000133808, {0x1400010bea0, 0x5, 0x5})
        github.com/spf13/cobra@v1.8.0/command.go:983 +0x840
github.com/spf13/cobra.(*Command).ExecuteC(0x140001ba008)
        github.com/spf13/cobra@v1.8.0/command.go:1115 +0x344
github.com/spf13/cobra.(*Command).Execute(...)
        github.com/spf13/cobra@v1.8.0/command.go:1039
github.com/spf13/cobra.(*Command).ExecuteContext(...)
        github.com/spf13/cobra@v1.8.0/command.go:1032
github.com/kong/deck/cmd.Execute({0x1040d8b10, 0x140003eeaf0})
        github.com/kong/deck/cmd/root.go:265 +0x64
main.main()
        github.com/kong/deck/main.go:31 +0x20

--- a/cmd/openapi2kong.go
+++ b/cmd/openapi2kong.go
@@ -69,11 +69,20 @@ func executeOpenapi2Kong(cmd *cobra.Command, _ []string) error {
 		}
 	}
 
+	var insoCompatibility bool
+	{
+		insoCompatibility, err = cmd.Flags().GetBool("inso-compatible")
+		if err != nil {
+			return fmt.Errorf("failed getting cli argument 'inso-compatible'; %w", err)
+		}
+	}
+
 	options := openapi2kong.O2kOptions{
 		Tags:                 entityTags,
 		DocName:              docName,
 		OIDC:                 generateSecurity,
 		IgnoreSecurityErrors: ignoreSecurityErrors,
+		InsoCompat:           insoCompatibility,
 	}
 
 	trackInfo := deckformat.HistoryNewEntry("openapi2kong")
@@ -127,4 +136,5 @@ directive from the file)`)
 	openapi2kongCmd.Flags().BoolP("generate-security", "", false, "generate OpenIDConnect plugins from the "+
 		"security directives")
 	openapi2kongCmd.Flags().BoolP("ignore-security-errors", "", false, "ignore errors for unsupported security schemes")
+	openapi2kongCmd.Flags().BoolP("inso-compatible", "", false, "generate the config in an Inso compatible way")
 }

--- a/openapi2kong/oas3_testfiles/09a-plugins-with-consumers.expected.json
+++ b/openapi2kong/oas3_testfiles/09a-plugins-with-consumers.expected.json
@@ -3,20 +3,6 @@
   "plugins": [
     {
       "config": {
-        "message": "The answer to life, the universe, and everything!",
-        "status_code": 403
-      },
-      "consumer": "johndoe2",
-      "id": "aa56031e-7155-599f-a9e9-93e6b271ba58",
-      "name": "request-termination",
-      "route": "simple-api-overview_uses-path-plugin",
-      "tags": [
-        "OAS3_import",
-        "OAS3file_09a-plugins-with-consumers.yaml"
-      ]
-    },
-    {
-      "config": {
         "message": "For a moment, nothing happened. Then, after a second or so, nothing continued to happen.",
         "status_code": 403
       },
@@ -24,6 +10,20 @@
       "id": "ead16074-ccb0-52dd-9f56-4193529e8ffa",
       "name": "request-termination",
       "route": "simple-api-overview_uses-ops-plugin",
+      "tags": [
+        "OAS3_import",
+        "OAS3file_09a-plugins-with-consumers.yaml"
+      ]
+    },
+    {
+      "config": {
+        "message": "The answer to life, the universe, and everything!",
+        "status_code": 403
+      },
+      "consumer": "johndoe2",
+      "id": "aa56031e-7155-599f-a9e9-93e6b271ba58",
+      "name": "request-termination",
+      "route": "simple-api-overview_uses-path-plugin",
       "tags": [
         "OAS3_import",
         "OAS3file_09a-plugins-with-consumers.yaml"

--- a/openapi2kong/openapi2kong.go
+++ b/openapi2kong/openapi2kong.go
@@ -530,10 +530,14 @@ func insertPlugin(list *[]*map[string]interface{}, newPlugin *map[string]interfa
 // getForeignKeyPlugins checks the pluginList for plugins that also have a foreign key
 // for a consumer, and moves them to the docPlugins array. Returns update docPlugins and pluginList.
 func getForeignKeyPlugins(
-	docPlugins *[]*map[string]interface{},
-	pluginList *[]*map[string]interface{},
-	foreignKey string, foreignValue string,
-) (*[]*map[string]interface{}, *[]*map[string]interface{}) {
+	docPlugins *[]*map[string]interface{}, // the current list of doc-level plugins (may be nil)
+	pluginList *[]*map[string]interface{}, // the list of entity-level plugins to check for foreign keys (may be nil)
+	foreignKey string, // the owner entity type: eg. "service", or "route"
+	foreignValue string, // the owner entity name/id: the value (service/route name)
+) (
+	*[]*map[string]interface{}, // updated slice of document level plugins
+	*[]*map[string]interface{}, // updated slice of entity level plugins
+) {
 	var genericPlugins []*map[string]interface{}
 	if docPlugins == nil {
 		genericPlugins = make([]*map[string]interface{}, 0)
@@ -1119,13 +1123,38 @@ func Convert(content []byte, opts O2kOptions) (map[string]interface{}, error) {
 	result["services"] = services
 	result["upstreams"] = upstreams
 	if len(*foreignKeyPlugins) > 0 {
+
+		// getSortKey returns a string that can be used to sort the plugins by name, service, route, and consumer (all
+		// the foreign keys that are possible).
+		getSortKey := func(p *map[string]interface{}) string {
+			plugin := *p
+			sep := string([]byte{0})
+			key := plugin["name"].(string) + sep
+
+			if plugin["service"] != nil {
+				key = key + plugin["service"].(string) + sep
+			} else {
+				key = key + sep
+			}
+
+			if plugin["route"] != nil {
+				key = key + plugin["route"].(string) + sep
+			} else {
+				key = key + sep
+			}
+
+			if plugin["consumer"] != nil {
+				key = key + plugin["consumer"].(string) + sep
+			} else {
+				key = key + sep
+			}
+
+			return key
+		}
+
 		sort.Slice(*foreignKeyPlugins,
 			func(i, j int) bool {
-				p1 := *(*foreignKeyPlugins)[i]
-				p2 := *(*foreignKeyPlugins)[j]
-				k1 := p1["name"].(string) + p1["id"].(string)
-				k2 := p2["name"].(string) + p2["id"].(string)
-				return k1 < k2
+				return getSortKey((*foreignKeyPlugins)[i]) < getSortKey((*foreignKeyPlugins)[j])
 			})
 		result["plugins"] = foreignKeyPlugins
 	}

